applications/ShallowWaterApplication/custom_utilities/replicate_model_part_utility.cpp

```diff
@@ -0,0 +1,238 @@
+
+    // Setting the nodal solution step variables
+    VariablesList& variables_list = mrOriginModelPart.GetNodalSolutionStepVariablesList();
+    mrDestinationModelPart.GetNodalSolutionStepVariablesList() = variables_list;
+
+    // In case we need to replicate the sub model parts
+    AssignUniqueModelPartCollectionTagUtility::IndexStringMapType collections;
+    AssignUniqueModelPartCollectionTagUtility::IndexIndexMapType origin_nodes_tags;
+    AssignUniqueModelPartCollectionTagUtility::IndexIndexMapType origin_elements_tags;
+    AssignUniqueModelPartCollectionTagUtility::IndexIndexMapType origin_conditions_tags;
+
+    std::unordered_map<IndexType, std::vector<IndexType>> dest_nodes_of_collection;
+    std::unordered_map<IndexType, std::vector<IndexType>> dest_elems_of_collection;
+    std::unordered_map<IndexType, std::vector<IndexType>> dest_conds_of_collection;
+
+    if (mReplicateSubModelParts)
+    {
+        AssignUniqueModelPartCollectionTagUtility(mrOriginModelPart).ComputeTags(
+            origin_nodes_tags,
+            origin_conditions_tags,
+            origin_elements_tags,
+            collections);
+    }
+
+    // Clone the nodes
+    mReplicatedNodes.clear();
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfNodes(); ++i)
+    {
+        auto it_node = mrOriginModelPart.NodesBegin() + i;
+        auto new_node = it_node->Clone();
+        new_node->SetId(++unique_node_id);
+        new_node->SetSolutionStepVariablesList(&variables_list);
+        mrDestinationModelPart.AddNode(new_node);
+        mReplicatedNodes[it_node->Id()] = new_node;
+        if (mReplicateSubModelParts)
+        {
+            auto tag = origin_nodes_tags[unique_node_id];
+            dest_nodes_of_collection[tag].push_back(unique_node_id);
+        }
+    }
+
+    // We also need to replicate the properties before creating the new elements
+    std::unordered_map<IndexType, Properties::Pointer> replicated_properties;
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfProperties(); ++i)
+    {
+        auto it_prop = mrOriginModelPart.PropertiesBegin() + i;
+        Properties::Pointer new_prop(new Properties(*it_prop));
+        new_prop->SetId(++unique_prop_id);
+        mrDestinationModelPart.AddProperties(new_prop);
+        replicated_properties[it_prop->Id()] = new_prop;
+    }
+
+    // Note: we need to manually create the geometry with the cloned nodes
+    for (IndexType i = 0; i < mrOriginModelPart.NumberOfElements(); ++i)
+    {
+        auto it_elem = mrOriginModelPart.ElementsBegin() + i;
+        PointerVector<Node<3>> new_elem_nodes;
+        for (auto& node : it_elem->GetGeometry())
+        {
+            new_elem_nodes.push_back(mReplicatedNodes[node.Id()]);
+        }
+        auto new_elem = it_elem->Clone(++unique_elem_id, new_elem_nodes);
+        new_elem->SetProperties(replicated_properties[it_elem->GetProperties().Id()]);
+        mrDestinationModelPart.AddElement(new_elem);
```

Comment:
It is faster to save the elements in an auxiliary array to add them altogether using the `AddElements`. Something like:
```
mrDestinationModelPart.AddElements(
    aux_array_with_element_pointers.begin(),
    aux_array_with_element_pointers.end());
```

Reply:
Done